Detectors/TRD/calibration/src/PadCalibCCDBBuilder.cxx

```diff
@@ -0,0 +1,551 @@
+  vCenter.push_back(hDet->GetNbinsX() / 2.);
+  vCenter.push_back(hDet->GetNbinsY() / 2.);
+
+  vPad1.push_back(hDet->GetXaxis()->GetBinCenter(coordinates[0]));
+  vPad1.push_back(hDet->GetYaxis()->GetBinCenter(coordinates[1]));
+
+  vPad2.push_back(hDet->GetXaxis()->GetBinCenter(coordinates[2]));
+  vPad2.push_back(hDet->GetYaxis()->GetBinCenter(coordinates[3]));
+
+  float dist1 = computeDistance(vPad1, vCenter);
+  float dist2 = computeDistance(vPad2, vCenter);
+
+  if ((dist1 < dist2) && ((hDet->GetBinContent(coordinates[2], coordinates[3]) - hDet->GetBinContent(coordinates[0], coordinates[1])) > allowedDifference)) {
+    replacePadCloserToCenter(hDet, coordinates[0], coordinates[1]);
+  }
+
+  if ((dist1 > dist2) && ((hDet->GetBinContent(coordinates[0], coordinates[1]) - hDet->GetBinContent(coordinates[2], coordinates[3])) > allowedDifference)) {
+    replacePadCloserToCenter(hDet, coordinates[2], coordinates[3]);
+  }
+}
+
+std::vector<int> PadCalibCCDBBuilder::compareGain(TH2F* hDet, int column, int row, int shiftcolumn, int shiftrow, float allowedDifference)
+{
+  std::vector<int> coordinates = {-1, -1, -1, -1};
+
+  int colMax = hDet->GetNbinsX();
+  int rowMax = hDet->GetNbinsY();
+
+  // only allow shift along one axis and only by one pad
+  if (!(shiftcolumn == 1 && shiftrow == 0) && !(shiftcolumn == 0 && shiftrow == 1))
+    return coordinates;
+
+  // cheks that the pad is valid
+  if ((column >= colMax) || (row >= rowMax))
+    return coordinates;
+
+  if ((column == colMax && shiftcolumn == 1) || (row == rowMax && shiftrow == 1))
+    return coordinates; // do not compare with overflow
+
+  float gain1 = hDet->GetBinContent(column, row);
+  float gain2 = hDet->GetBinContent(column + shiftcolumn, row + shiftrow);
+
+  if ((gain1 == 0 && gain2 > 0) || (gain1 > 0 && gain2 == 0) || (abs(gain1 - gain2) > allowedDifference)) {
+    coordinates[0] = column;
+    coordinates[1] = row;
+    coordinates[2] = column + shiftcolumn;
+    coordinates[3] = row + shiftrow;
+  }
+
+  return coordinates;
+}
+
+float PadCalibCCDBBuilder::computeDetectorAverage(TH2F* hDet)
+{ // computes an average over filled cells
+  // cells are accessed through their values, not bin coordinates
+  // the average os computed over absolute values as inter/extrapolated cells
+  // have negatiove values
+
+  float average = 0.;
+  int nBinsUsed = 0;
+  for (int irow = 0; irow < hDet->GetNbinsY(); irow++) {
+    for (int icol = 0; icol < hDet->GetNbinsX(); icol++) {
+      float currentGain = TMath::Abs(hDet->GetBinContent(icol + 1, irow + 1));
+      if (currentGain == 0)
+        continue;
+      average += currentGain;
+      nBinsUsed++;
+    }
+  }
+  average /= nBinsUsed;
+
+  return average;
+}
+
+float PadCalibCCDBBuilder::computeDistance(std::vector<float> pad1, std::vector<float> pad2)
+{
+  float distance = -1.;
+
+  if (pad1.size() != pad2.size()) {
+    std::cerr << "Something fishy with the pad coordinates!" << std::endl;
```

Comment:
Is using this for logging okay in calibration code or is LOG(error) the way to go here? @tdietel @martenole